tests/driver_lis2dh12/main.c

```diff
@@ -11,170 +11,699 @@
+};
 
 /* allocate device descriptor */
 static lis2dh12_t dev;
 
 #ifdef MODULE_LIS2DH12_INT
-/* control interrupt */
-typedef struct {
-    uint8_t line;
-    mutex_t *lock;
-    uint8_t *flags;
-} lis_ctx;
-
-/* timer lock */
-static uint8_t isr_flags;
-static mutex_t isr_mtx = MUTEX_INIT_LOCKED;
-static lis_ctx ctx[2] = {
-    {
-        .line  = 1,
-        .lock  = &isr_mtx,
-        .flags = &isr_flags,
-    }, {
-        .line  = 2,
-        .lock  = &isr_mtx,
-        .flags = &isr_flags,
-    }
-};
+/* Interrupt lines */
+static uint8_t line1 = 1;
+static uint8_t line2 = 2;
 
-/* interrupt callback function. */
-static void lis2dh12_int_cb(void* _ctx) {
-    lis_ctx *control = _ctx;
+/* Interrupt params */
+static lis2dh12_int_params_t params_int1 = {0};
+static lis2dh12_int_params_t params_int2 = {0};
 
-    *control->flags |= control->line;
+/* Interrupt source register */
+static uint8_t int1_src;
+#endif /* MODULE_LIS2DH12_INT */
 
-    mutex_unlock(control->lock);
-}
+/* FIFO data memory */
+static lis2dh12_fifo_data_t data_fifo[NUM_FIFO_VALUES];
+/* FIFO configuration */
+static lis2dh12_fifo_t fifo_cfg = {
+    .FIFO_set_INT2 = false,
+    .FIFO_watermark = 10,
+    .FIFO_mode = LIS2DH12_FIFO_MODE_STREAMtoFIFO,
+};
 
-/* print interrupt register */
-static void lis2dh12_int_reg_content(lis2dh12_t *dev, uint8_t pin){
+/* Memory to print current data */
+static char str_out[3][8];
 
-    assert(pin == LIS2DH12_INT1 || pin == LIS2DH12_INT2);
+/* current lis acceleration data */
+static int16_t data_lis[3];
 
-    uint8_t buffer;
-    lis2dh12_read_int_src(dev, &buffer, pin);
+/* highpass configuration */
+lis2dh12_highpass_t highpass_cfg = {
+    .Highpass_mode = LIS2DH12_HP_MODE_REFERENCE,
+    .Highpass_freq = LIS2DH12_HP_FREQ_DIV100,
+    .CLICK_enable = false,
+    .INT1_enable = false,
+    .INT2_enable = false,
+    .DATA_OUT_enable = false,
+};
 
-    printf("content SRC_Reg_%d: 0x%02x\n", pin, buffer);
-    printf("\t XL %d\n", !!(buffer & LIS2DH12_INT_SRC_XL));
-    printf("\t XH %d\n", !!(buffer & LIS2DH12_INT_SRC_XH));
-    printf("\t YL %d\n", !!(buffer & LIS2DH12_INT_SRC_YL));
-    printf("\t YH %d\n", !!(buffer & LIS2DH12_INT_SRC_YH));
-    printf("\t ZL %d\n", !!(buffer & LIS2DH12_INT_SRC_ZL));
-    printf("\t ZH %d\n", !!(buffer & LIS2DH12_INT_SRC_ZH));
-    printf("\t IA %d\n", !!(buffer & LIS2DH12_INT_SRC_IA));
-}
-#endif
+/* reference data */
+static uint8_t reference_value;
 
-int main(void)
-{
+/* shock threshold */
+static int16_t shock_thold;
 
 #ifdef MODULE_LIS2DH12_INT
-    uint8_t flags = 0;
-#endif
+/* previous values */
+static int16_t old_data_lis[3];
+static uint8_t int1_src_old;
+
+/* lis2dh12 interrupt callback function. */
+static void lis2dh12_int_cb(void* l) {
+
+    /* disable IRQ until lis_process is done */
+    gpio_irq_disable(dev.p->int1_pin);
+    gpio_irq_disable(dev.p->int2_pin);
+
+    /* reset click source */
+    lis2dh12_read_click_src(&dev, &click_src_reg);
```

Comment:
You can't do I2C communication in Interrupt context

Reply:
> Why no I2C communication

Because there is no way to make sure that the interrupt does not interrupt a currently ongoing I2C transfer done by the application. 
Usually this is prevented by mutex, but we can't block on a mutex in the interrupt.

If the flag needs to be cleared we should rethink the API and not expose the raw interrupt callback to the application.

Reply:
Ah, I thought we have some interrupt priorities and the I2C can not be interrupted by the LIS. Are there a method in RIOT to set the priority of the I2C higher than the interrupt pins? I think this should fix this.

Reply:
Interrupts can always interrupt code in thread mode, not the other way round.
The only way would be to disable the interrupt, but that would be an ugly hack. 

Let's stay within the API contract and not perform any I2C / SPI communication in interrupt context.